atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -0,0 +1,428 @@
+ *   ... this is one of the reasons why there is a new system.
+ **/
+public class CassandraClientPool {
+    private static final Logger log = LoggerFactory.getLogger(CassandraClientPool.class);
+    private static long BANNED_HOST_BACKOFF_TIME_MS = TimeUnit.MINUTES.toMillis(5);
+    private static int MAX_TRIES = 3;
+
+    RangeMap<LightweightOPPToken, List<InetSocketAddress>> tokenMap = ImmutableRangeMap.of();
+    Map<InetSocketAddress, Long> blacklistedHosts = Maps.newHashMap();
+    Map<InetSocketAddress, CassandraClientPoolingContainer> currentPools = Maps.newHashMap();
+    CassandraKeyValueServiceConfig config;
+    ScheduledThreadPoolExecutor refreshDaemon;
+
+
+    private class LightweightOPPToken implements Comparable<LightweightOPPToken> {
+        byte[] bytes;
+
+        public LightweightOPPToken(byte[] bytes) {
+            this.bytes = bytes;
+        }
+
+        @Override
+        public int compareTo(LightweightOPPToken other) {
+            return UnsignedBytes.lexicographicalComparator().compare(this.bytes, other.bytes);
+        }
+    }
+
+    public CassandraClientPool(CassandraKeyValueServiceConfig config) {
+        this.config = config;
+        refreshDaemon = PTExecutors.newScheduledThreadPool(1, new ThreadFactoryBuilder().setDaemon(true).setNameFormat("CassandraClientPoolingContainerRefreshDaemon-%d").build());
+        refreshDaemon.scheduleWithFixedDelay(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    refreshPool();
+                } catch (Throwable t) {
+                    log.error("Failed to refresh Cassandra KVS pool. Extended periods of being unable to refresh will cause perf degradation.", t);
+                }
+            }
+        }, CassandraConstants.SECONDS_BETWEEN_GETTING_HOST_LIST, CassandraConstants.SECONDS_BETWEEN_GETTING_HOST_LIST, TimeUnit.SECONDS);
+
+        refreshPool(); // ensure we've initialized before returning
+    }
+
+
+    public void shutdown() {
+        refreshDaemon.shutdown();
+        currentPools.forEach((address, cassandraClientPoolingContainer) -> cassandraClientPoolingContainer.shutdownPooling());
+    }
+
+    private void refreshPool() {
+        checkAndUpdateBlacklist();
+
+        Set<InetSocketAddress> serversToAdd = config.servers();
+        Set<InetSocketAddress> serversToRemove = ImmutableSet.of();
+
+        if (config.autoRefreshNodes()) {
+            refreshTokenRanges(); // re-use token mapping as list of hosts in the cluster
+            for (List<InetSocketAddress> rangeOwners : tokenMap.asMapOfRanges().values()) {
+                for (InetSocketAddress address : rangeOwners) {
+                    serversToAdd.add(address);
+                }
+            }
+        }
+        Sets.difference(Sets.difference(serversToAdd, currentPools.keySet()), blacklistedHosts.keySet());
```

Comment:
Think you want to assign the result of this as it doesn't modify the sets, it just returns a lazy view that you need to assign to use.